### PR TITLE
[optim] prevent problematic tensor aliasing in lr_scheduler

### LINKED ISSUE #162359: SequentialLR aliases lr Tensor with initial_lr, corrupting base_lrs for chained schedulers
### 🐛 Describe the bug

1. When we use an optimizer with a Tensor learning rate (as [recommended by the docs to avoid recompilation](https://docs.pytorch.org/tutorials/recipes/compiling_optimizer_lr_scheduler.html)), `SequentialLR.__init__` assigns `group["lr"] = group["initial_lr"]`, aliasing the tensors.
2. The initial step of the first scheduler writes the lr in-place, which also mutates the `initial_lr` and all chained schedulers' `base_lrs` that reference it.
3. Chained schedulers which rely on `base_lrs` therefore start from the wrong base (e.g. scaled by `start_factor` if we used a `LinearLR` first).

For example, if we use a Tensor learning rate and `SequentialLR` with `schedulers=[LinearLR, CosineAnnealingLR]`, the `base_lrs` of our chained schedulers (and thus our `CosineAnnealingLR` outputs) are scaled by the `LinearLR`'s `start_factor`:

Image

`LinearLR.get_lr` does not rely on `base_lrs`, hence the jump at the milestone. With the bug fixed, we obtain the intended schedule:

Image

Here's a reproduction:
```python
import torch
from torch import optim
import matplotlib.pyplot as plt

# 1. Use a tensor learning rate. The choice of optimizer doesn't matter.
lr = 1.0
x = torch.tensor(0.0, device='meta')
opt = optim.AdamW([x], lr=torch.tensor(lr))

# 2. Initialize our chained schedulers.
milestone, total_steps = 40, 100
start_factor, end_factor = 0.2, 1.0
warmup = optim.lr_scheduler.LinearLR(opt, start_factor, end_factor, total_iters=milestone)
decay = optim.lr_scheduler.CosineAnnealingLR(opt, T_max=total_steps-milestone)

# Note that each scheduler's base_lr aliases with the optimizer's initial_lr.
assert warmup.base_lrs[0].is_set_to(opt.param_groups[0]['initial_lr'])
assert decay.base_lrs[0].is_set_to(opt.param_groups[0]['initial_lr'])

# 3. Initialize our SequentialLR.
scheduler = optim.lr_scheduler.SequentialLR(opt, schedulers=[warmup, decay], milestones=[milestone])

# SequentialLR.__init__ aliases our optimizer's lr and initial_lr.
# for group in self.optimizer.param_groups:
#     group["lr"] = group["initial_lr"]
assert opt.param_groups[0]['lr'].is_set_to(opt.param_groups[0]['initial_lr'])

# Which means they're also aliased with the base_lrs of each scheduler!
assert (warmup.base_lrs[0].data_ptr() 
    == decay.base_lrs[0].data_ptr()
    == opt.param_groups[0]['initial_lr'].data_ptr() 
    == opt.param_groups[0]['lr'].data_ptr())

# It also calls _initial_step() on our LinearLR, setting all tensors to start_factor * lr.
assert (start_factor * lr
    == opt.param_groups[0]['lr']
    == opt.param_groups[0]['initial_lr']
    == warmup.base_lrs[0]
    == decay.base_lrs[0])

# 4. Now if we do something which breaks the aliases, e.g. save/load the optimizer's state_dict...
opt.load_state_dict(opt.state_dict())
assert not warmup.base_lrs[0].data_ptr() == opt.param_groups[0]['initial_lr'].data_ptr()

# Our base_lrs are incorrect. The LinearLR will work as intended (since get_lr updates based on the
# previous iteration, not base_lr), but the outputs of further schedulers are incorrectly scaled by 
# start_factor. In our case, these are the outputs of CosineAnnealingLR which appear after the milestone.
lrs = []
for step in range(total_steps):
    scheduler.step()
    lrs.append(scheduler.get_last_lr()[0].item())

plt.plot(lrs)
plt.show()
```

### Versions

PyTorch version: 2.8.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: version 4.1.1
Libc version: N/A

Python version: 3.13.7 (main, Aug 14 2025, 11:12:11) [Clang 17.0.0 (clang-1700.0.13.3)] (64-bit runtime)
Python platform: macOS-15.6.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] No relevant packages

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### LINKED ISSUE #163093: [lr_scheduler] `ReduceLROnPlateau._reduce_lr` triggers recompilation with tensor lr
### 🐛 Describe the bug

`ReduceLROnPlateau._reduce_lr` sets `param_group["lr"]` to a float, [triggering recompilation](https://docs.pytorch.org/tutorials/recipes/compiling_optimizer_lr_scheduler.html) if the optimizer uses a tensor LR:
```python
for i, param_group in enumerate(self.optimizer.param_groups):
    old_lr = float(param_group["lr"])
    new_lr = max(old_lr * self.factor, self.min_lrs[i])
    if old_lr - new_lr > self.eps:
        param_group["lr"] = new_lr
```

Here's a reproduction:
```python
import torch
from torch import optim, nn

torch._logging.set_logs(recompiles=True)
device = "cuda"
model = nn.Linear(1, 1).to(device)

lr = torch.tensor(0.1)
opt = optim.Adam(model.parameters(), lr=lr)
scheduler = optim.lr_scheduler.ReduceLROnPlateau(opt, factor=0.5, patience=1, min_lr=0.001)

# From https://docs.pytorch.org/tutorials/recipes/compiling_optimizer_lr_scheduler.html
@torch.compile(fullgraph=False)
def fn(metric):
    opt.step()
    scheduler.step(metric)

total_steps = 8
fake_metrics = torch.linspace(1.0, 0.0, total_steps, device=device)
fake_metrics[3:6] = fake_metrics[3]

for metric in fake_metrics:
    fn(metric)

# W0912 20:46:20.358000 89262 torch/_logging/_internal.py:1199] [0/0] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
# V0912 20:46:20.564000 89262 torch/_dynamo/guards.py:4295] [2/1] [__recompiles] Recompiling function step in /workspace/pytorch/torch/optim/adam.py:213
# V0912 20:46:20.564000 89262 torch/_dynamo/guards.py:4295] [2/1] [__recompiles]     triggered by the following guard failure(s):
# V0912 20:46:20.564000 89262 torch/_dynamo/guards.py:4295] [2/1] [__recompiles]     - 2/0: expected type of 'self.param_groups[0]['lr']' to be a tensor type, ' but found <class 'float'>
```

### Versions

PyTorch version: 2.10.0.dev20250916
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: version 3.25.0
Libc version: N/A

Python version: 3.13.7 (main, Aug 14 2025, 11:12:11) [Clang 17.0.0 (clang-1700.0.13.3)] (64-bit runtime)
Python platform: macOS-15.6.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] flake8==7.3.0
[pip3] flake8-bugbear==24.12.12
[pip3] flake8-comprehensions==3.16.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==2024.24.12
[pip3] flake8-pyi==25.5.0
[pip3] flake8_simplify==0.22.0
[pip3] mypy==1.16.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.1.0
[pip3] optree==0.13.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by filipviz**
Prevents edge cases in SequentialLR and ReduceLROnPlateau which could corrupt learning rates or trigger recompilation.

Supersedes #162360
Fixes #162359
Fixes #163093

While putting #162360 together, I noticed the class of issue I was fixing (i.e. unintended aliasing in lr_schedulers when using Tensor lrs) appeared in several other places. @janeyx99 suggested I put together a follow-up PR.

There are several bugs resembling the one fixed in #162360. I added a helper to fix these:
```python
def _update_param_group_val(param_group: dict[str, Any], key: str, val: float | Tensor):
    """Set param_group[key] to val without aliasing or assignment when they're both tensors.
    Raises a KeyError if param_group[key] does not exist.
    """
    if isinstance(param_group[key], Tensor):
        param_group[key].fill_(_to_scalar(val))
    else:
        param_group[key] = val
```

And applied it to fix bugs in `SequentialLR.__init__` and `LRScheduler._update_lr`. I also added it to `CyclicLR.__init__` which was using an equivalent pattern, and `CosineAnnealingWarmRestarts.step` which *should* have had a similar issue:
```python
for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
    param_group["lr"] = lr
```

But did not, because `get_lr()` actually returns tensors when using a tensor lr (despite its `list[float]` return type annotation). Relying on this propagation seems fragile, so I conservatively added the method here as well. I'll be fixing the type annotations and several related issues in followup PRs built off of this one.

### CODE REVIEW

**Review by janeyx99 — APPROVED** [review]

**Comment by janeyx99** [comment]
@pytorchbot merge

thanks for your detailed work!

**Comment by filipviz** [comment]
Sorry! I accidentally pushed 39c342e24f5aedca7dd1b6acdd65bbf0d21dbf36 (which was supposed to be a follow-up PR) to this branch. Let me know how you'd like me to handle this @janeyx99. That PR fixes a number of type annotations and an issue where `LRScheduler._update_lr`, `ChainedScheduler.__init__`, `ChainedScheduler.step`, `ReduceLROnPlateau.__init__`, `ReduceLROOnPlateau.step`, and `CosineAnnealingWarmRestarts.step` previously aliased `self._last_lr` with `group["lr"]` if `group["lr"]` was a tensor, due to patterns like this:
```python
self._last_lr: list[float] = [group["lr"] for group in self.optimizer.param_groups]
```

I added another helper to address this:
```python
def _param_groups_val_list(optimizer: Optimizer, key: str) -> list[Any]:
    """Create a list containing group[key] for each optimizer param_group.
    Prevents aliasing when group[key] could be a Tensor.
    Raises a KeyError when group[key] does not exist.
    """
    return [
        group[key].clone() if isinstance(group[key], Tensor) else group[key]
        for group in optimizer.param_groups
    ]
```

**Comment by janeyx99** [comment]
Let's undo that change and land the original first!

I can review the followup PR separately. It's good to keep changes minimal per PR.

**Comment by filipviz** [comment]
Same page @janeyx99. Reverted that commit but added `test_reduce_lr_on_plateau_preserves_lr_type` back in (since this PR fixes that bug). Let me know if you'd like any other changes.

**Review by janeyx99 — APPROVED** [review]
This commit also fixes CosineAnnealingWarmRestarts, yes?

**Comment by janeyx99** [comment]
@pytorchbot merge

no need for other changes as long as CI is green!
